### PR TITLE
[RAISE-BP] Allow more `arith` operations as `tt.addptr` operands

### PR DESCRIPTION
**Pull request by victor-eds**
Allow `arith.[addi|constant|muli]` as `tt.addptr` operands when raising.

### CODE REVIEW

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 1]
```diff
@@ -222,84 +267,30 @@ struct TritonRaiseBlockPointer
...
+        .Default([](Operation *op) {
+          llvm::dbgs() << "TritonRaiseBlockPointer: encountered addptr operand "
+                          "produced by an unsupported operation\n"
+                       << op << "\n";
+          return failure();
+        });
   }
 
-  LogicalResult visitOperandMakeRange(triton::MakeRangeOp rangeOp,
-                                      PtrState &state, Location loc,
```
Comment (victor-eds) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:241]: Changed with:

```c++
template <>
LogicalResult visitAddPointerOperand(triton::MakeRangeOp
```

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 2]
```diff
@@ -222,84 +267,30 @@ struct TritonRaiseBlockPointer
...
-        builder.create<arith::ConstantIntOp>(loc, 0, shapeAndStridesWidth));
-    state.sizes.push_back(shape[0]);
-
-    LLVM_DEBUG(llvm::dbgs() << "MakeRange state: " << state << "\n";);
-
-    return success();
-  }
-
-  LogicalResult visitOperandSplat(triton::SplatOp splatOp, PtrState &state,
-                                  Location loc, OpBuilder &builder) {
```
Comment (victor-eds) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:267]: Changed with:

```c++
template <>
LogicalResult visitAddPointerOperand(triton::SplatOp
```

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 3]
```diff
@@ -349,4 +339,143 @@ struct TritonRaiseBlockPointer
   llvm::SmallDenseMap<Value, PtrState> knownPtrs;
   IRMapping ptrMap;
 };
+
+template <>
+LogicalResult
+TritonRaiseBlockPointer::visitAddPointerOperand(triton::MakeRangeOp rangeOp,
```
Comment (mfrancepillois) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:345]: It seems to me that
`template <> 
LogicalResult visitOperand(...)`
is probably a better name for this function than
`template <> 
LogicalResult 
visitAddPointerOperand(...)`
which suggest that it is a visitor to the `triton::AddPtrOp` operation.

**Review by mfrancepillois — APPROVED** [review]

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp (hunk 3 above)
Comment (victor-eds) [reply third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:345]: It's to visit `tt.addptr` operands only, not other operation operands, hence the name. I don't think it's bad...

File: test/Triton/raise-block-pointer.mlir [hunk 4]
```diff
@@ -68,3 +68,90 @@ tt.func @test_addptr_splat_splat_2d(%arg0 : !tt.ptr<f32>, %arg1: i64, %arg2: ten
...
+  %0 = tt.splat %arg0 : !tt.ptr<f32> -> tensor<128x!tt.ptr<f32>>
+  %1 = tt.splat %arg1 : i32 -> tensor<128xi32>
+  %2 = tt.make_range {start = 0 : i32, end = 128 : i32} : tensor<128xi32>
+  %3 = arith.muli %1, %2 : tensor<128xi32>
+  %4 = tt.addptr %0, %3 : tensor<128x!tt.ptr<f32>>, tensor<128xi32>
+  %5 = tt.load %4 : tensor<128x!tt.ptr<f32>>
+  tt.return %5 : tensor<128xf32>
+}
+
+// CHECK-LABEL:   tt.func @test_addptr_splat_const(
```
Comment (etiotto) [inline test/Triton/raise-block-pointer.mlir:129]: [nit]: `test_addptr_splat_const` -> `test_const_splat_addptr`

File: test/Triton/raise-block-pointer.mlir [hunk 5]
```diff
@@ -68,3 +68,90 @@ tt.func @test_addptr_splat_splat_2d(%arg0 : !tt.ptr<f32>, %arg1: i64, %arg2: ten
...
+// CHECK:           tt.return %[[VAL_4]] : tensor<128xf32>
+tt.func @test_addptr_splat_const(%arg0 : !tt.ptr<f32>) -> tensor<128xf32> {
+  %cst = arith.constant dense<512> : tensor<128xi32>
+  %0 = tt.splat %arg0 : !tt.ptr<f32> -> tensor<128x!tt.ptr<f32>>
+  %1 = tt.addptr %0, %cst : tensor<128x!tt.ptr<f32>>, tensor<128xi32>
+  %2 = tt.load %1 : tensor<128x!tt.ptr<f32>>
+  tt.return %2 : tensor<128xf32>
+}
+
+// CHECK-LABEL:   tt.func @test_addptr_splat_const_2d(
```
Comment (etiotto) [inline test/Triton/raise-block-pointer.mlir:144]: ditto

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 6]
```diff
@@ -223,115 +269,215 @@ struct TritonRaiseBlockPointer
...
 
-    llvm::errs() << "TritonRaiseBlockPointer: encountered addptr operand "
-                    "produced by an unsupported operation\n"
-                 << operand << "\n";
-
-    return failure();
+    return TypeSwitch<Operation *, LogicalResult>(definingOp)
+        .Case<arith::AddIOp, arith::ConstantOp, arith::MulIOp,
+              triton::MakeRangeOp, triton::SplatOp>(
+            [this, &state, loc, &builder](auto op) {
```
Comment (etiotto) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:282]: [nit]: just capture with "&"?

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 7]
```diff
@@ -223,115 +269,215 @@ struct TritonRaiseBlockPointer
...
   llvm::SmallDenseMap<Value, PtrState> knownPtrs;
   IRMapping ptrMap;
 };
+
+template <>
+LogicalResult
+TritonRaiseBlockPointer::visitAddPointerOperand(triton::MakeRangeOp rangeOp,
+                                                PtrState &state, Location loc,
+                                                OpBuilder &builder) {
+  assert(state.isEmpty());
```
Comment (etiotto) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:350]: Add assert msg

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 8]
```diff
@@ -223,115 +269,215 @@ struct TritonRaiseBlockPointer
...
+
+  return success();
+}
+
+template <>
+LogicalResult
+TritonRaiseBlockPointer::visitAddPointerOperand(triton::SplatOp splatOp,
+                                                PtrState &state, Location loc,
+                                                OpBuilder &builder) {
+  assert(state.isEmpty());
```
Comment (etiotto) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:378]: Add assert msg

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 9]
```diff
@@ -223,115 +269,215 @@ struct TritonRaiseBlockPointer
...
+
+  LLVM_DEBUG(llvm::dbgs() << "Splat state: " << state << "\n";);
+
+  return success();
+}
+
+template <>
+LogicalResult TritonRaiseBlockPointer::visitAddPointerOperand(
+    arith::AddIOp addOp, PtrState &state, Location loc, OpBuilder &builder) {
+  assert(state.isEmpty());
```
Comment (etiotto) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:417]: Add assert msg, also below

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp [hunk 10]
```diff
@@ -223,115 +269,215 @@ struct TritonRaiseBlockPointer
...
+}
+
+template <>
+LogicalResult TritonRaiseBlockPointer::visitAddPointerOperand(
+    arith::ConstantOp op, PtrState &state, Location loc, OpBuilder &builder) {
+  assert(state.isEmpty());
+
+  auto attr = cast<DenseElementsAttr>(op.getValue());
+  Type elementType = attr.getElementType();
+  assert(attr.isSplat() && isa<IntegerType>(elementType));
```
Comment (etiotto) [inline third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:463]: Add assert msg

**Review by etiotto — APPROVED** [review]

File: third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp (hunk 6 above)
Comment (victor-eds) [reply third_party/intel/lib/TritonRaiseBlockPointer/TritonRaiseBlockPointer.cpp:282]: I prefer it this way if you don't mind much, but I can change

**Pull request merged**